File: packages/health_connector_hk_ios/test/unit_tests/src/mappers/health_record_mappers/blood_pressure/blood_pressure_record_mapper_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hk_ios/src/mappers/health_record_mappers/blood_pressure/blood_pressure_record_mapper.dart';
import 'package:health_connector_hk_ios/src/pigeon/health_connector_hk_ios_api.g.dart';

import '../../../../utils/fake_data.dart';

void main() {
  group(
    'BloodPressureRecordMapper',
    () {
      group(
        'BloodPressureRecordToDto',
        () {
          test(
            'converts BloodPressureRecord to BloodPressureRecordDto',
            () {
              final time = FakeData.fakeTime;

              final record = BloodPressureRecord(
                id: HealthRecordId(FakeData.fakeId),
                time: time,
                zoneOffsetSeconds: FakeData.fakeTime.timeZoneOffset.inSeconds,
                metadata: Metadata.internal(
                  dataOrigin: const DataOrigin(FakeData.fakeDataOrigin),
                  recordingMethod: RecordingMethod.manualEntry,
                  clientRecordVersion: 1,
                  device: const Device(type: DeviceType.phone),
                ),
                systolic: const Pressure.millimetersOfMercury(120.0),
                diastolic: const Pressure.millimetersOfMercury(80.0),
              );

              final dto = record.toDto();

              expect(dto.id, FakeData.fakeId);
              expect(dto.time, time.millisecondsSinceEpoch);
              expect(
                dto.zoneOffsetSeconds,
                FakeData.fakeTime.timeZoneOffset.inSeconds,
              );
              expect(dto.systolicInMillimetersOfMercury, 120.0);
              expect(dto.diastolicInMillimetersOfMercury, 80.0);
              expect(dto.metadata.dataOrigin, FakeData.fakeDataOrigin);
            },
          );
        },
      );

      group(
        'BloodPressureRecordDtoToDomain',
        () {
          test(
            'converts BloodPressureRecordDto to BloodPressureRecord',
            () {
              final time = FakeData.fakeTime;

              final dto = BloodPressureRecordDto(
                id: FakeData.fakeId,
                time: time.millisecondsSinceEpoch,
                zoneOffsetSeconds: FakeData.fakeTime.timeZoneOffset.inSeconds,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.activelyRecorded,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                systolicInMillimetersOfMercury: 125.0,
                diastolicInMillimetersOfMercury: 85.0,
                measurementLocation: MeasurementLocationDto.unknown,
                bodyPosition: BodyPositionDto.unknown,
              );

              final record = dto.toDomain();

              expect(record.id.value, FakeData.fakeId);
              expect(record.time, time);
              expect(record.systolic.inMillimetersOfMercury, 125.0);
              expect(record.diastolic.inMillimetersOfMercury, 85.0);
              expect(
                record.metadata.dataOrigin?.packageName,
                FakeData.fakeDataOrigin,
              );
            },
          );

          test(
            'converts BloodPressureRecordDto with null id to '
            'domain with none id',
            () {
              final time = FakeData.fakeTime;

              final dto = BloodPressureRecordDto(
                time: time.millisecondsSinceEpoch,
                zoneOffsetSeconds: FakeData.fakeTime.timeZoneOffset.inSeconds,
                metadata: MetadataDto(
                  dataOrigin: FakeData.fakeDataOrigin,
                  recordingMethod: RecordingMethodDto.manualEntry,
                  clientRecordVersion: 1,
                  deviceType: DeviceTypeDto.phone,
                ),
                systolicInMillimetersOfMercury: 115.0,
                diastolicInMillimetersOfMercury: 75.0,
                measurementLocation: MeasurementLocationDto.unknown,
                bodyPosition: BodyPositionDto.unknown,
              );

              final record = dto.toDomain();

              expect(record.id, HealthRecordId.none);
            },
          );
        },
      );
    },
  );
}

File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_record_mappers/BloodPressureRecordMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_record_mappers

import androidx.health.connect.client.records.BloodPressureRecord
import androidx.health.connect.client.records.metadata.Metadata
import androidx.health.connect.client.units.Pressure
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.BloodPressureRecordDto
import com.phamtunglam.health_connector_hc_android.pigeon.BodyPositionDto
import com.phamtunglam.health_connector_hc_android.pigeon.DeviceTypeDto
import com.phamtunglam.health_connector_hc_android.pigeon.MeasurementLocationDto
import com.phamtunglam.health_connector_hc_android.pigeon.MetadataDto
import com.phamtunglam.health_connector_hc_android.pigeon.RecordingMethodDto
import io.kotest.matchers.shouldBe
import java.time.Instant
import java.time.ZoneOffset
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("BloodPressureRecordMapper")
class BloodPressureRecordMapperTest {

    private companion object {
        const val TEST_ID = "blood-pressure-id"
        const val TEST_TIME = 1609459200000L
        const val TEST_SYSTOLIC = 120.0
        const val TEST_DIASTOLIC = 80.0
        val TEST_ZONE_OFFSET: ZoneOffset = ZoneOffset.ofHours(1)
    }

    @Test
    @DisplayName("GIVEN BloodPressureRecord → WHEN toDto → THEN converts correctly")
    fun whenRecordToDto_thenConvertsCorrectly() {
        // Given
        val record = BloodPressureRecord(
            systolic = Pressure.millimetersOfMercury(TEST_SYSTOLIC),
            diastolic = Pressure.millimetersOfMercury(TEST_DIASTOLIC),
            bodyPosition = BloodPressureRecord.BODY_POSITION_SITTING_DOWN,
            measurementLocation = BloodPressureRecord.MEASUREMENT_LOCATION_LEFT_UPPER_ARM,
            time = Instant.ofEpochMilli(TEST_TIME),
            zoneOffset = TEST_ZONE_OFFSET,
            metadata = Metadata.manualEntry(),
        )

        // When
        val result = record.toDto()

        // Then
        result.systolicInMillimetersOfMercury shouldBe TEST_SYSTOLIC
        result.diastolicInMillimetersOfMercury shouldBe TEST_DIASTOLIC
        result.bodyPosition shouldBe BodyPositionDto.SITTING_DOWN
        result.measurementLocation shouldBe MeasurementLocationDto.LEFT_UPPER_ARM
        result.time shouldBe TEST_TIME
        result.zoneOffsetSeconds shouldBe TEST_ZONE_OFFSET.totalSeconds.toLong()
    }

    @Test
    @DisplayName("GIVEN BloodPressureRecordDto → WHEN toHealthConnect → THEN creates correctly")
    fun whenDtoToHealthConnect_thenCreatesCorrectly() {
        // Given
        val dto = BloodPressureRecordDto(
            id = TEST_ID,
            time = TEST_TIME,
            zoneOffsetSeconds = TEST_ZONE_OFFSET.totalSeconds.toLong(),
            systolicInMillimetersOfMercury = TEST_SYSTOLIC,
            diastolicInMillimetersOfMercury = TEST_DIASTOLIC,
            bodyPosition = BodyPositionDto.SITTING_DOWN,
            measurementLocation = MeasurementLocationDto.LEFT_UPPER_ARM,
            metadata = MetadataDto(
                dataOrigin = "com.example.app",
                recordingMethod = RecordingMethodDto.MANUAL_ENTRY,
                deviceType = DeviceTypeDto.PHONE,
            ),
        )

        // When
        val result = dto.toHealthConnect()

        // Then
        result.systolic.inMillimetersOfMercury shouldBe TEST_SYSTOLIC
        result.diastolic.inMillimetersOfMercury shouldBe TEST_DIASTOLIC
        result.bodyPosition shouldBe BloodPressureRecord.BODY_POSITION_SITTING_DOWN
        result.measurementLocation shouldBe BloodPressureRecord.MEASUREMENT_LOCATION_LEFT_UPPER_ARM
        result.time shouldBe Instant.ofEpochMilli(TEST_TIME)
        result.zoneOffset shouldBe TEST_ZONE_OFFSET
        result.metadata.id shouldBe TEST_ID
    }
}
